### DISCUSSION TITLE
Some questions about the new Luau support

### DISCUSSION THREAD

**Discussion by setzer22** [mlua-rs/mlua#136] [Q&A] [answered]
Congratulations on the new release :tada: The new Luau support looks very nice!

I read about some of your comments in #59 and I'd like to ask a bit more about it:

- You mentioned userdata has to be emulated because it doesn't exist in Luau. How is this userdata emulation performed? Is there going to be a significant performance hit when using userdata with Luau?
- You also mention requires had to be emulated: Is there no `require` support in luau? And in that case, did you add some sort of Rust function to replace it?
- What are the other adaptations you had to make? I'm very curious about the differences between Luau and Lua but I couldn't find much on their documentation.

*EDIT:* I just tried downloading the new version and it seems to work, but I tried annotating some functions and I can't seem to trigger any type errors or warnings. What's the workflow to ensure the extra type checks in Luau are exposed to the user?

**Comment by khvzak** [comment, marked as answer]
> How is this userdata emulation performed? Is there going to be a significant performance hit when using userdata with Luau?

Not userdata itself, but associated uservalues are emulated using the registry and weak tables.
Uservalue is a custom data attached to userdata via [`set_user_value`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html#method.set_user_value).
[This is how it's emulated](https://github.com/khvzak/mlua/blob/master/src/ffi/luau/compat.rs#L225)
It's definitely slower than in Lua 5.4, but still in my opinion very nice feature to have.

>  Is there no require support in luau?

Yes, Luau does not support modules out of the box (as I see the Luau code).
They emulated require in the CLI repl program [here](https://github.com/Roblox/luau/blob/master/CLI/Repl.cpp#L83)

>  And in that case, did you add some sort of Rust function to replace it?

Yep, the [function](https://github.com/khvzak/mlua/blob/master/src/lua.rs#L2631) I added

> What are the other adaptations you had to make? 

It's vanilla Luau. Major difference is lack of `__gc` support for userdata. Fortunately they provide an alternative using a new api function `lua_newuserdatadtor` that allows to attach destructor. mlua heavily relies on destructors and it required some effort to adopt new approach.
And Luau does not support hooks.

>  but I tried annotating some functions and I can't seem to trigger any type errors or warnings. What's the workflow to ensure the extra type checks in Luau are exposed to the user?

As I see, Luau compiler ignores any type annotations and they needs to be checked first by Luau analyzer. You can download binaries [here](https://github.com/Roblox/luau/releases)


Also Luau supports `vector(x, y, z)` native datatype that I'll add soon to mlua. And I don't configure "sandboxed" Lua environment that have some negative performance impact (sandboxed env is faster than non-sandboxed). I'll add sandboxing support to a next release after checking how it works.

**Comment by setzer22** [comment]
Thanks for the answers! I am working on an extension layer for [my project](https://github.com/setzer22/blackjack), and so far `mlua` looks like the best Lua integration crate. I was very happy to see Luau support land right after I started using the library, keep up the good work!

> Not userdata itself, but associated uservalues

Ah! I didn't even know about uservalues. Thanks for clearing that up.

> As I see, Luau compiler ignores any type annotations and they needs to be checked first by Luau analyzer.

Would it be possible to include the analyzer in `mlua` so that you could do something like: `lua.analyze(...)` before calling `lua.load(...)` to get a list of errors and display that to the user? I don't think it's a big deal if it's too hard, since once can always shell out to luau-analyze, but it would be nice to have it all in the same library :)

> Also Luau supports vector(x, y, z) native datatype 

*Very* nice! I was just asking about some lightweight way to share vectors between rust and lua. While tables work, I am super interested in having something more performant that doesn't require allocations. The main use case for the Lua layer in my project is lua callbacks that get vectors, do some math, and return vectors, so that is my number 1 bottleneck.

>  I'll add sandboxing support to a next release after checking how it works.

Definitely interested on that too! I'll be keeping an eye on future updates :smile:

**Reply by khvzak** [reply]
> Would it be possible to include the analyzer in mlua so that you could do something like: lua.analyze(...) before calling lua.load(...) to get a list of errors and display that to the user? 

Sorry, probably not now. Analyzer is quite independent thing that even does not has C api. Without analyzer you still can write normal dynamically-typed Lua code.

> Very nice! I was just asking about some lightweight way to share vectors between rust and lua. While tables work, I am super interested in having something more performant that doesn't require allocations. The main use case for the Lua layer in my project is lua callbacks that get vectors, do some math, and return vectors, so that is my number 1 bottleneck.

I released v0.8.0-beta.2 that adds [`Value::Vector`](https://docs.rs/mlua/0.8.0-beta.2/mlua/enum.Value.html#variant.Vector) datatype  support for Luau. It should out of the box support native arithmenic operations according the https://luau-lang.org/performance#native-vector-math

**Comment by cortesi** [comment]
I've just released luau-analyze - simple, clean bindings to the luau analyzer. 

https://github.com/cortesi/luau-analyze
